lib/ansible/modules/mount_facts.py

```diff
@@ -0,0 +1,637 @@
+    - /etc/mnttab: 5 columns (mnttab[4] must contain UNIX timestamp)
+    - /etc/fstab: 4-6 columns (fstab[4] is optional and historically 0-9, but can be any int)
+    - /etc/filesystems: multi-line, not column-based, and specific to AIX
+    """
+    seen = set()
+    for file in sources:
+        if file in seen:
+            continue
+        seen.add(file)
+
+        if not (lines := get_file_content(file, "").splitlines()):
+            continue
+
+        for gen_mounts in [gen_vfstab_entries, gen_mnttab_entries, gen_fstab_entries, gen_aix_filesystems_entries]:
+            with suppress(IndexError, ValueError):
+                # mypy type broken before Python 3.11, only works if either
+                # * the list of functions excludes gen_aix_filesystems_entries
+                # * the list of functions only contains gen_aix_filesystems_entries
+                yield from [
+                    (file, mount_info.mount_point, mount_info.line, mount_info.fields) for mount_info in gen_mounts(lines)  # type: ignore[attr-defined]
+                ]
+                break
+
+
+def get_sources(module: AnsibleModule) -> list[str]:
+    """Return a list of filenames from the requested sources."""
+    sources: list[str] = []
+    for source in module.params["sources"] or ["all"]:
+        if not source:
+            module.fail_json(msg="sources contains an empty string")
+
+        if source in {"dynamic", "all"}:
+            sources.extend(DYNAMIC_SOURCES)
+            if module.params["mount_binary"] is not None:
+                sources.append(module.params["mount_binary"])
+        if source in {"static", "all"}:
+            sources.extend(STATIC_SOURCES)
+
+        elif source not in {"static", "dynamic", "all"}:
+            sources.append(source)
+    return sources
+
+
+def gen_mounts_by_source(module: AnsibleModule):
+    """Iterate over the sources and yield tuples containing the source, mount point, source line(s), and the parsed result."""
+    sources = get_sources(module)
+
+    if len(set(sources)) < len(sources):
+        module.warn(f"mount_facts option 'sources' contains duplicate entries, repeat sources will be ignored: {sources}")
+
+    mount_binary = module.params["mount_binary"]
+    for source in sources:
+        if source == mount_binary:
+            stdout = run_mount_bin(module, mount_binary)
+            yield from [(mount_binary, *astuple(mount_info)) for mount_info in gen_mounts_from_stdout(stdout)]
+            continue
+
+        seen = set()
+        for mount in gen_mounts_by_file([source]):
+            mount_point = mount[1]
+
+            if module.params["include_aggregate_mounts"] is None:
+                if mount_point not in seen:
+                    seen.add(mount_point)
+                else:
+                    # Warn about duplicate mounts within the same source since there may be multiple devices mounted.
+                    module.warn(f"Ignoring repeat mounts for {mount_point}. "
```

Comment:
It'd be good to include which module sourced the warning (especially when running with other stuff under `gather_facts`), eg:

```suggestion
                    module.warn(f'mount_facts: ignoring repeat mounts for {mount_point}. "
```

Also (not a merge blocker), but it might be good to collapse to a single warning that lists them all to cut down on the boilerplate parts.

Reply:
I committed your suggestion, but I will add an update to collapse it to a single warning.

Reply:
I ran into an issue with symlinked sources, so if it's not a blocker, I'm going to leave this as-is. For example /etc/mtab is symlinked to /proc/mounts and since the module collects from both, if I include the source in the warning, the same warning for the same mount points is given twice. Here's what I was attempting:

```python
    duplicates_msg = ""
    # FIXME: would need to deduplicate by os.path.realpath(src)
    for src, mnts in file_source_mount_points.items():
        if len(set(mnts)) == len(mnts):
            continue
        duplicates = [mnt for mnt in mnts if mnts.count(mnt) > 1]
        duplicates_msg += f"{src}: {', '.join(duplicates)}\n"

    if module.params["include_aggregate_mounts"] is None and duplicates_msg:
        module.warn(f"mount_facts: ignoring repeat mounts for the following sources with duplicates:\n{duplicates_msg}"
                    "You can disable this warning by configuring the 'include_aggregate_mounts' option as True or False.")

```

Reply:
I collapsed this to a single warning in the latest commit, and fixed the symlinks issue so the aggregate_mounts don't contain duplicates.